sugar_test.go

```diff
@@ -31,33 +31,88 @@ import (
+		}
+	}
+
 	tests := []struct {
+		desc     string
 		args     []interface{}
 		expected []zapcore.Field
 		errLogs  []observer.LoggedEntry
 	}{
-		{nil, []zapcore.Field{}, nil},
-		{[]interface{}{}, []zapcore.Field{}, nil},
-		{[]interface{}{"should ignore"}, []zapcore.Field{}, []observer.LoggedEntry{ignored}},
-		{[]interface{}{"foo", 42, "true", "bar"}, []zapcore.Field{Int("foo", 42), String("true", "bar")}, nil},
-		{[]interface{}{Int("foo", 42)}, []zapcore.Field{Int("foo", 42)}, nil},
-		{[]interface{}{Int("foo", 42), "should ignore"}, []zapcore.Field{Int("foo", 42)}, []observer.LoggedEntry{ignored}},
 		{
-			[]interface{}{"first", "field", Int("foo", 42), "baz", "quux", "should ignore"},
-			[]zapcore.Field{String("first", "field"), Int("foo", 42), String("baz", "quux")},
-			[]observer.LoggedEntry{ignored},
+			desc:     "nil args",
+			args:     nil,
+			expected: []zapcore.Field{},
+			errLogs:  nil,
 		},
 		{
-			args:     []interface{}{"foo", 42, true, "bar"},
+			desc:     "empty slice of args",
+			args:     []interface{}{},
+			expected: []zapcore.Field{},
+			errLogs:  nil,
+		},
+		{
+			desc:     "just a dangling key",
+			args:     []interface{}{"should ignore"},
+			expected: []zapcore.Field{},
+			errLogs:  []observer.LoggedEntry{ignored("should ignore")},
+		},
+		{
+			desc:     "well-formed key-value pairs",
+			args:     []interface{}{"foo", 42, "true", "bar"},
 			expected: []zapcore.Field{Int("foo", 42), String("true", "bar")},
-			errLogs: []observer.LoggedEntry{{
-				Entry:   zapcore.Entry{Level: DPanicLevel, Message: _nonStringKeyErrMsg},
-				Context: []zapcore.Field{Int("position", 2), Any("key", true), Any("value", "bar")},
-			}},
+			errLogs:  nil,
+		},
+		{
+			desc:     "just a structured field",
+			args:     []interface{}{Int("foo", 42)},
+			expected: []zapcore.Field{Int("foo", 42)},
+			errLogs:  nil,
+		},
+		{
+			desc:     "structured field and a dangling key",
+			args:     []interface{}{Int("foo", 42), "dangling"},
+			expected: []zapcore.Field{Int("foo", 42)},
+			errLogs:  []observer.LoggedEntry{ignored("dangling")},
+		},
+		{
+			desc:     "key-value pair and a dangling key",
+			args:     []interface{}{"foo", 42, "dangling"},
+			expected: []zapcore.Field{Int("foo", 42)},
+			errLogs:  []observer.LoggedEntry{ignored("dangling")},
+		},
+		{
+			desc:     "pairs, a structured field, and a dangling key",
+			args:     []interface{}{"first", "field", Int("foo", 42), "baz", "quux", "dangling"},
+			expected: []zapcore.Field{String("first", "field"), Int("foo", 42), String("baz", "quux")},
+			errLogs:  []observer.LoggedEntry{ignored("dangling")},
+		},
+		{
+			desc:     "one non-string key",
+			args:     []interface{}{"foo", 42, true, "bar"},
+			expected: []zapcore.Field{Int("foo", 42)},
+			errLogs:  []observer.LoggedEntry{nonString(invalidPair{2, true, "bar"})},
+		},
+		{
+			desc:     "pairs, structured fields, non-string keys, and a dangling key",
```

Comment:
other cases that occur to me:
- dangling non-string value
- other things that aren't strings as keys, like structs, interface values, etc
- more interleaving, like `pair(s), structured field(s), pair(s), structured field(s)`

I hesitate to suggest programmatically generating test cases, especially in a world where we still can't use sub-tests (I dearly look forward to a post go 1.6 support world ;-) ).